### PR TITLE
Bottle

### PR DESCRIPTION
**Pull request by JeremyRand**
Use Bottle for the HTTP server, implement a REST API using Bottle.

### CODE REVIEW

**Comment by JeremyRand** [comment]
This was successfully tested by @josephbisch .  https://forum.namecoin.info/viewtopic.php?p=14598#p14598

Merging in 72 hours unless there are objections.  @phelixbtc

**Comment by phelixbtc** [comment]
This breaks, maybe a missing "import traceback":
http://127.0.0.2:8080/data/ASDFgetData/d/nameid.json

How would you add calling the http GUI?

**Comment by JeremyRand** [comment]
@phelixbtc Fixed that error, thanks for noticing.

I would suggest implementing the HTTP GUI by adding a route for '/<plugin>/<method>/<args:path>.html', and letting plugins supply methods that output HTML rather than JSON.  That way it stays modular.  I'm open to alternative suggestions if you have something better.

File: plugin/pluginHttp.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            result = methodRpc(method, *params)
+        except AttributeError, e:
+            if app['debug']: traceback.print_exc()
+            raise Exception('Method "' + method + '" not supported by plugin "' + plugin + '"')
+        #except Exception, e:
+        #    if app['debug']: traceback.print_exc()
+        #    return template('Exception : ' + str(e))
+    
+        # restore stdout
+        sys.stdout = sys.__stdout__
```
Comment (phelixbtc) [inline plugin/pluginHttp.py]: should be commented out?

Comment (JeremyRand) [inline plugin/pluginHttp.py]: That exception is caught a few lines later: see https://github.com/namecoin/nmcontrol/pull/67/files#diff-f38c2f8e6722eee6a9af231bc0c8afa5R99 .  The original RPC plugin converted exceptions to a True value in the section you're citing, which isn't very Pythonic, so I changed it to simply pass on exceptions to a later point, when the response is being generated.  The intent is (in future) to produce error codes for various exceptions rather than always having an error code of 1.

Comment (phelixbtc) [inline plugin/pluginHttp.py]: I meant it seems that stdout is never changed above.

Comment (JeremyRand) [inline plugin/pluginHttp.py]: Oh.  You may be right, but that code is copied verbatim from the RPC plugin.  I think anything that changes that should be a separate PR.

**Comment by phelixbtc** [comment]
ACK

**Comment by JeremyRand** [comment]
Thanks, merging.
